library/ssl_cli.c

```diff
@@ -3485,6 +3489,162 @@ static int ssl_write_client_key_exchange( mbedtls_ssl_context *ssl )
           MBEDTLS_KEY_EXCHANGE_ECDHE_ECDSA_ENABLED ||
           MBEDTLS_KEY_EXCHANGE_ECDH_RSA_ENABLED ||
           MBEDTLS_KEY_EXCHANGE_ECDH_ECDSA_ENABLED */
+#if defined(MBEDTLS_USE_PSA_CRYPTO) &&                           \
+    defined(MBEDTLS_KEY_EXCHANGE_ECDHE_PSK_ENABLED)
+    if( ciphersuite_info->key_exchange == MBEDTLS_KEY_EXCHANGE_ECDHE_PSK )
+    {
+        psa_status_t status = PSA_ERROR_CORRUPTION_DETECTED;
+        psa_status_t destruction_status = PSA_ERROR_CORRUPTION_DETECTED;
+        psa_key_attributes_t key_attributes;
+
+        mbedtls_ssl_handshake_params *handshake = ssl->handshake;
+
+        /*
+         * opaque psk_identity<0..2^16-1>;
+         */
+        if( ssl_conf_has_static_psk( ssl->conf ) == 0 )
+            /* We don't offer PSK suites if we don't have a PSK,
+             * and we check that the server's choice is among the
+             * ciphersuites we offered, so this should never happen. */
+            return( MBEDTLS_ERR_SSL_INTERNAL_ERROR );
+
+        /* Opaque PSKs are currently only supported for PSK-only suites. */
+        if( ssl_conf_has_static_raw_psk( ssl->conf ) == 0 )
+            return( MBEDTLS_ERR_SSL_FEATURE_UNAVAILABLE );
+
+        /* uint16 to store content length */
+        const size_t content_len_size = 2;
+
+        header_len = 4;
+        content_len = ssl->conf->psk_identity_len;
+
+        if( header_len + content_len_size + content_len
+                    > MBEDTLS_SSL_OUT_CONTENT_LEN )
+        {
+            MBEDTLS_SSL_DEBUG_MSG( 1,
+                ( "psk identity too long or SSL buffer too short" ) );
+            return( MBEDTLS_ERR_SSL_BUFFER_TOO_SMALL );
+        }
+
+        ssl->out_msg[header_len++] = MBEDTLS_BYTE_1( content_len );
+        ssl->out_msg[header_len++] = MBEDTLS_BYTE_0( content_len );
```

Comment:
I know this is copied from existing code, but I think mutating `header_len` is not good for readability: the variable name really does not suggest that. IMO it's a bit worse here than in the existing code, just because the variable is used more.

I suggest declaring `unsigned char *p = ssl->out_msg + header_len` then updating that - we have a fairly consistent convention throughout our code that `p` suggests that kind of usage. (Simply renaming `header_len` to `current_offset` would be consistent too, but less aligned with the rest of the code base. Hmm, now that I think of it, we sometimes use `n` to mean `current_offset` though I think that's less common.)

Edit after reading further: OK, so there's already a `p` pointer in the second part of the function (computing/assembling the PMS). While it's generally not a good idea to use the same variable for different things in the same function, in that case I think it's OK, as the function really has two fairly distinct parts. Alternatively, we could use `p` in one part and `q` in the other. Or perhaps that's an argument for using `out_msg[current_offset]` in the first part after all. At least it makes it explicit what general buffer we're writing to.